docs/CONTRIBUTING.md

```diff
@@ -0,0 +1,246 @@
+but you are welcome to pick any unassigned issue
+
+#### Creating an Issue
+- Create the issue from the TA Projects Board (linked above)
+  - Click "add item" at the bottom of the `Todo` list
+  - Type `#autograder` to connect the issue to the Autograder
+  - Type a title and hit enter
+- Use a clear, descriptive title that uses one of these prefixes:
+  - `Backend:` server side changes
+  - `Frontend:` client (web) side changes
+  - `Fullstack:` for issues that require front and backend changes
+  - `Docs:` changes to documentation (either markdown files or JSDocs/JavaDocs)
+  - `Tests:` adding or modifying tests
+  - `Dev:` changes to enhance the Autograder development experience
+- Provide enough context so others can understand the purpose
+- Add labels as appropriate ([Here's the list](https://github.com/softwareconstruction240/autograder/issues/labels))
+- Link related issues
+- Issues should be narrow in scope
+
+#### Working on an Issue
+- Assign the issue to yourself before beginning work (this is how we know 
+  who is working on what, and gives you a sort of ownership over the issue)
+- Create a new branch ([more info here](#2-branches))
+- If you find a new issue, file it separately (don't mix concerns)
+- Leave a comment from time to time with updates on your progress
+
+### 2. Branches
+Branches in git allow us to all work on separate things in the same codebase.
+You will do all your work in a branch.
+
+#### Creating a branch
+Create the branch directly from the issue so your work is linked.
+
+- Always branch from `main` 
+  - (exceptions exist if you know what you're doing)
+- Use descriptive names that reflect what you're working on:
+    - `add-late-submission`
+    - `admin-display-verification-status`
+    - `server-communicator`
+    - `extract-service-logic-from-controllers`
+- Use kebab-case (lowercase with hyphens)
+- Keep names concise but clear
+- One branch per issue (don't mix different features/fixes)
+
+### 3. Making Changes
+Now that you have a branch, you can start to code! Hurray! 🎉
+Here are important guidelines to follow:
+
+- Review and follow the Style Guide (_Coming Soon_)
+- Write tests for your code. All new features and functionality must have tests demonstrating correctness
+  - Update existing tests if you change functionality
+  - _New contributors: A great way to better understand the codebase is to try writing tests for existing code_
+- Keep commits focused and meaningful
+  - Each commit should represent one logical change
+  - Write clear commit messages that describe that change ([See guidelines](#writing-good-commit-messages))
+- Update documentation as you go
+  - Add JSDoc/JavaDoc comments to new code
+    - _New contributors: Another great way to learn the codebase is to document existing code_
+  - Update repo Markdown docs if needed
+  - Code should for the most part be self-documenting. If not, make sure to add inline comments
```

Comment:
This is a contested topic, and I'm not sure any "right" answer exists. Sometimes, there are reasons to include comments, and sometimes that's not the best way to move forward. However, as you tried to argue recently in softwareconstruction240/checkstyle#3, sometimes the "most correct" way is not the way to move forward in the present moment.

The following summary of the "code comments are bad" debate highlights that there is a debate. Trying to decide one "correct" answer in a contested arena is going to be difficult.

## Rundown of Key Arguments from Both Sides of the Debate
> Credit: This section was kindly produced by [ChatGPT](https://chatgpt.com/share/672c723a-b5bc-800c-9af4-ee7dd9302e61).

Here’s a rundown of arguments from both sides of the debate on whether inline comments are a code smell or a readability enhancer:

### Arguments that Inline Comments are a Code Smell

1. **Code Should Be Self-Explanatory**  
   Advocates of this view believe that code itself should be clean and readable enough that it doesn't require inline comments. If comments are necessary to explain code, it may indicate the code is too complex or poorly structured.

2. **Comments Can Become Outdated**  
   Code is often updated, but comments may not be maintained with the same rigor, leading to situations where comments misrepresent the code. This inconsistency can confuse developers, especially in collaborative environments.

3. **Comments Might Encourage Bad Practices**  
   Relying on comments to explain complex logic can discourage developers from refactoring the code to make it simpler. Inline comments can sometimes mask bad design choices rather than addressing the underlying issues.

4. **Brevity and Focus**  
   Reducing the number of comments can keep code more focused and to the point, avoiding distractions. Some believe that excessive comments make it harder to quickly understand the main structure and purpose of the code.

### Arguments that Inline Comments Enhance Readability

1. **Contextual Clarification**  
   Inline comments can provide context or explain why certain decisions were made, especially when dealing with intricate or non-obvious logic, external dependencies, or unusual design choices that are not immediately clear.

2. **Onboarding and Collaboration**  
   For new team members or contributors unfamiliar with a codebase, inline comments can significantly reduce the learning curve by explaining complex sections. This can be particularly helpful in large teams or open-source projects.

3. **Code Can’t Capture Intent Fully**  
   Even well-written code may not always convey the intent behind it. Inline comments can clarify the purpose behind a line or block, helping future developers understand the reasoning rather than just the mechanics.

4. **Highlighting Workarounds or Known Issues**  
   Sometimes, developers implement workarounds for known issues or bugs. Inline comments can explain these so future developers aren’t left puzzled by non-standard implementations or don’t accidentally remove necessary code quirks.

Ultimately, whether inline comments are considered a code smell often depends on the codebase, team culture, and specific project requirements.

Reply:
I'm not the biggest fan of inline comments, and the principle should be self documenting code. But I think there's definitely some places where it's not realistic. The way this section was worded was my attempt to encourage self documenting code, but require comments where it's not possible or difficult (In particular, I'm thinking of CSS, but this could happen other places). I'll go ahead and try to rework this until an official decision is made.

Reply:
I get that there's an argument, I just don't think the CONTRIBUTING.md or this PR is the place. Hence the suggestion to link to the style guide.

I don't feel strongly here, especially since we decided to prioritize simplicity and brevity for new devs, so I'd be fine with a slight rewording or even just leaving it as it is.